saleor/webhook/subscription_payload.py

```diff
@@ -0,0 +1,116 @@
+from typing import Any, Dict, Optional
+
+from celery.utils.log import get_task_logger
+from django.core.handlers.base import BaseHandler
+from django.http import HttpRequest
+from django.test.client import RequestFactory
+from graphene_django.settings import graphene_settings
+from graphene_django.views import instantiate_middleware
+from graphql import get_default_backend, parse
+from graphql.error import GraphQLSyntaxError
+from graphql.execution.middleware import MiddlewareManager
+from promise import Promise
+
+from ..app.models import App
+
+logger = get_task_logger(__name__)
+
+
+def validate_subscription_query(query: str) -> bool:
+    from ..graphql.api import schema
+
+    graphql_backend = get_default_backend()
+    try:
+        graphql_backend.document_from_string(schema, query)
+    except (ValueError, GraphQLSyntaxError):
+        return False
+    return True
+
+
+def initialize_context() -> HttpRequest:
+    """Prepare a request object for webhook subscription.
+
+    It creates a dummy request object and initialize middleware on it. It is required
+    to process a request in the same way as API logic does.
+    return: HttpRequest
+    """
+    handler = BaseHandler()
+    context = RequestFactory().request()
+    handler.load_middleware()
+    handler.get_response(context)
+    return context
+
+
+def generate_payload_from_subscription(
+    event_type: str,
+    subscribable_object,
+    subscription_query: Optional[str],
+    context: HttpRequest,
+    app: Optional[App] = None,
+) -> Optional[Dict[str, Any]]:
+    """Generate webhook payload from subscription query.
+
+    It uses a graphql's engine to build payload by using the same logic as response.
+    As an input it expects given event type and object and the query which will be
+    used to resolve a payload.
+    event_type: is a event which will be triggered.
+    subscribable_object: is a object which have a dedicated own type in Subscription
+    definition.
+    subscription_query: query used to prepare a payload via graphql engine.
+    context: A dummy request used to share context between apps in order to use
+    dataloaders benefits.
+    app: the owner of the given payload. Required in case when webhook contains
+    protected fields.
+    return: A payload ready to send via webhook. None if the function was not able to
+    generate a payload
+    """
+    from ..graphql.api import schema
+
+    graphql_backend = get_default_backend()
+    ast = parse(subscription_query)  # type: ignore
+    document = graphql_backend.document_from_string(
+        schema,
+        ast,
+    )
+    graphql_middleware = graphene_settings.MIDDLEWARE
+    graphql_middleware = MiddlewareManager(
+        instantiate_middleware(graphql_middleware), wrap_in_promise=False
+    )
+    app_id = app.pk if app else None
+
+    context.app = app  # type: ignore
+
+    results = document.execute(
```

Comment:
@fowczarek  can you confirm? Should we somehow define the replica that we want to use it here?

Reply:
Nope. Replicas are not related to GraphQLView.